File: flutter-wrapper/test/proxy/proxy_runtime_test.dart
import 'dart:async';

import 'package:flutter_test/flutter_test.dart';
import 'package:h5_app/proxy/proxy_runtime.dart';

void main() {
  test('isHealthy flips to false when supervisor gives up', () async {
    final gaveUp = Completer<void>();
    var healthy = true;

    // The runtime exposes a hook for tests to install a gaveUp future and a
    // health setter callback. The hook is `attachGaveUp(Future<void>, void Function(bool))`.
    ProxyRuntime.instance.attachGaveUp(gaveUp.future, (h) => healthy = h);
    gaveUp.complete();
    await Future<void>.delayed(Duration.zero);

    expect(healthy, isFalse);
  });
}

File: electron-wrapper/test/proxy/proxy-runtime.test.js
import { describe, it, expect } from 'vitest';
import { ProxyRuntime } from '../../proxy/proxy-runtime.js';

describe('ProxyRuntime gaveUp wiring', () => {
  it('flips isHealthy=false when supervisor gives up', async () => {
    const rt = new ProxyRuntime();
    let resolveGaveUp;
    const gaveUp = new Promise(r => { resolveGaveUp = r; });
    rt.isHealthy = true;
    rt._timer = setInterval(() => {}, 1_000_000);
    rt.attachGaveUp({ gaveUp });
    resolveGaveUp();
    await new Promise(r => setImmediate(r));
    expect(rt.isHealthy).toBe(false);
    expect(rt._timer).toBeNull();
  });
});
